File: poc3/core/kernel.py
"""Proof kernel: sequent calculus with 10 rules."""


# === Types ===

class Mem:
    def __init__(self, left, right):
        self.left = left
        self.right = right

class Neg:
    def __init__(self, operand):
        self.operand = operand

class Implies:
    def __init__(self, left, right):
        self.left = left
        self.right = right

class Forall:
    def __init__(self, var, body):
        self.var = var
        self.body = body

class Sequent:
    def __init__(self, left, right):
        _check_set(left)
        _check_set(right)
        self.left = left
        self.right = right

def _check_set(lst):
    for i, a in enumerate(lst):
        for j in range(i + 1, len(lst)):
            if same(a, lst[j]):
                raise ValueError(f'sequent: duplicate formula')

class Proof:
    def __init__(self, sequent):
        self.sequent = sequent


# === Substitution ===

def subst(formula, old, new):
    if isinstance(formula, str):
        return new if formula == old else formula
    if isinstance(formula, Mem):
        return Mem(subst(formula.left, old, new), subst(formula.right, old, new))
    if isinstance(formula, Neg):
        return Neg(subst(formula.operand, old, new))
    if isinstance(formula, Implies):
        return Implies(subst(formula.left, old, new), subst(formula.right, old, new))
    if isinstance(formula, Forall):
        if formula.var == old:
            return formula
        return Forall(formula.var, subst(formula.body, old, new))
    raise ValueError(f'cannot subst: {type(formula).__name__}')


# === Alpha-equivalence ===

def same(a, b):
    return eq(a, b, [])

def eq(a, b, env):
    if isinstance(a, str) and isinstance(b, str):
        for v1, v2 in env:
            if a == v1: return b == v2
            if b == v2: return False
        return a == b
    if type(a) is not type(b):
        return False
    if isinstance(a, Mem):
        return eq(a.left, b.left, env) and eq(a.right, b.right, env)
    if isinstance(a, Neg):
        return eq(a.operand, b.operand, env)
    if isinstance(a, Implies):
        return eq(a.left, b.left, env) and eq(a.right, b.right, env)
    if isinstance(a, Forall):
        return eq(a.body, b.body, [(a.var, b.var)] + env)
    return False


# === Free variables ===

def free_vars(formula, bound=None):
    if bound is None:
        bound = set()
    if isinstance(formula, str):
        return {formula} - bound
    if isinstance(formula, Mem):
        return free_vars(formula.left, bound) | free_vars(formula.right, bound)
    if isinstance(formula, Neg):
        return free_vars(formula.operand, bound)
    if isinstance(formula, Implies):
        return free_vars(formula.left, bound) | free_vars(formula.right, bound)
    if isinstance(formula, Forall):
        return free_vars(formula.body, bound | {formula.var})
    return set()


def bound_vars(formula):
    if isinstance(formula, str):
        return set()
    if isinstance(formula, Mem):
        return bound_vars(formula.left) | bound_vars(formula.right)
    if isinstance(formula, Neg):
        return bound_vars(formula.operand)
    if isinstance(formula, Implies):
        return bound_vars(formula.left) | bound_vars(formula.right)
    if isinstance(formula, Forall):
        return {formula.var} | bound_vars(formula.body)
    return set()


# === Helpers ===

def fin(f, lst):
    return any(same(f, g) for g in lst)

def remove(lst, f):
    result = []
    removed = False
    for g in lst:
        if not removed and same(f, g):
            removed = True
        else:
            result.append(g)
    return result

def set_add(lst, f):
    if fin(f, lst):
        return list(lst)
    return list(lst) + [f]

def eq_sequent(a, b):
    return is_permutation(a.left, b.left) and is_permutation(a.right, b.right)

def is_permutation(a, b):
    if len(a) != len(b):
        return False
    used = [False] * len(b)
    for f in a:
        found = False
        for j, g in enumerate(b):
            if not used[j] and same(f, g):
                used[j] = True
                found = True
                break
        if not found:
            return False
    return True


# === Proof rules ===

def proof(seq, rule, premises, principal, term=None):
    if not check_rule(seq, rule, premises, principal, term):
        raise ValueError(f'invalid proof step: {rule}')
    return Proof(seq)


def check_rule(s, rule, premises, principal, term):
    ps = [p.sequent for p in premises]

    match rule:
        # A in G, A in D  =>  G |- D
        case "axiom":
            return (len(ps) == 0
                    and fin(principal, s.left) and fin(principal, s.right))
        # G |- D, A  =>  G, ~A |- D
        case "neg_left":
            return (len(ps) == 1 and isinstance(principal, Neg)
                    and fin(principal, s.left)
                    and eq_sequent(ps[0], Sequent(
                        remove(s.left, principal),
                        set_add(s.right, principal.operand))))
        # A, G |- D  =>  G |- D, ~A
        case "neg_right":
            return (len(ps) == 1 and isinstance(principal, Neg)
                    and fin(principal, s.right)
                    and eq_sequent(ps[0], Sequent(
                        set_add(s.left, principal.operand),
                        remove(s.right, principal))))
        # G |- D, A    B, G |- D  =>  G, A->B |- D
        case "implies_left":
            if len(ps) != 2 or not isinstance(principal, Implies):
                return False
            if not fin(principal, s.left):
                return False
            G = remove(s.left, principal)
            return (eq_sequent(ps[0], Sequent(G, set_add(s.right, principal.left)))
                    and eq_sequent(ps[1], Sequent(set_add(G, principal.right), s.right)))
        # A, G |- D, B  =>  G |- D, A->B
        case "implies_right":
            if len(ps) != 1 or not isinstance(principal, Implies):
                return False
            if not fin(principal, s.right):
                return False
            D = remove(s.right, principal)
            return eq_sequent(ps[0], Sequent(
                set_add(s.left, principal.left), set_add(D, principal.right)))
        # p[t/x], G |- D  =>  G, Ax.p |- D   (t is var, no capture)
        case "forall_left":
            if len(ps) != 1 or not isinstance(term, str) or not isinstance(principal, Forall):
                return False
            if not fin(principal, s.left):
                return False
            if term in bound_vars(principal.body):
                return False
            G = remove(s.left, principal)
            substituted = subst(principal.body, principal.var, term)
            return eq_sequent(ps[0], Sequent(set_add(G, substituted), s.right))
        # G |- D, p[t/x]  =>  G |- D, Ax.p   (t eigenvariable: not free in G,D, no capture)
        case "forall_right":
            if len(ps) != 1 or not isinstance(term, str) or not isinstance(principal, Forall):
                return False
            if not fin(principal, s.right):
                return False
            if term in bound_vars(principal.body):
                return False
            D = remove(s.right, principal)
            if any(term in free_vars(f) for f in s.left + D):
                return False
            substituted = subst(principal.body, principal.var, term)
            return eq_sequent(ps[0], Sequent(s.left, set_add(D, substituted)))
        # G |- D, A    A, G |- D  =>  G |- D
        case "cut":
            return (len(ps) == 2
                    and eq_sequent(ps[0], Sequent(s.left, set_add(s.right, principal)))
                    and eq_sequent(ps[1], Sequent(set_add(s.left, principal), s.right)))
        # G |- D  =>  G, A |- D
        case "weakening_left":
            if len(ps) != 1:
                return False
            if not fin(principal, s.left):
                return False
            if fin(principal, ps[0].left):
                return eq_sequent(ps[0], s)
            return eq_sequent(ps[0], Sequent(remove(s.left, principal), s.right))
        # G |- D  =>  G |- D, A
        case "weakening_right":
            if len(ps) != 1:
                return False
            if not fin(principal, s.right):
                return False
            if fin(principal, ps[0].right):
                return eq_sequent(ps[0], s)
            return eq_sequent(ps[0], Sequent(s.left, remove(s.right, principal)))
    return False


# === ZFC pattern matching ===

class Capture:
    def __init__(self, name):
        self.name = name

ANY = Capture('_')

def pmatch(pattern, formula, bindings=None):
    if bindings is None:
        bindings = {}
    if isinstance(pattern, Capture):
        bindings[pattern.name] = formula
        return True
    if isinstance(pattern, str) and isinstance(formula, str):
        if pattern in bindings:
            return bindings[pattern] == formula
        bindings[pattern] = formula
        return True
    if type(pattern) is not type(formula):
        return False
    if isinstance(pattern, Mem):
        return pmatch(pattern.left, formula.left, bindings) and pmatch(pattern.right, formula.right, bindings)
    if isinstance(pattern, Neg):
        return pmatch(pattern.operand, formula.operand, bindings)
    if isinstance(pattern, Implies):
        return pmatch(pattern.left, formula.left, bindings) and pmatch(pattern.right, formula.right, bindings)
    if isinstance(pattern, Forall):
        return pmatch(pattern.var, formula.var, bindings) and pmatch(pattern.body, formula.body, bindings)
    return False

# Derived helpers for building patterns
def PAnd(a, b): return Neg(Implies(a, Neg(b)))
def POr(a, b): return Implies(Neg(a), b)
def PIff(a, b): return Neg(Implies(Implies(a, b), Neg(Implies(b, a))))
def PExists(v, body): return Neg(Forall(v, Neg(body)))
def PEqv(a, b, z="_z"): return Forall(z, PIff(Mem(z, a), Mem(z, b)))

ZFC = [
    # Extensionality
    Forall("x", Forall("y", Implies(
        Forall("z", PIff(Mem("z", "x"), Mem("z", "y"))),
        Forall("z", PIff(Mem("x", "z"), Mem("y", "z")))))),
    # EmptySet
    PExists("b", Forall("x", Neg(Mem("x", "b")))),
    # Pairing
    Forall("x", Forall("y", PExists("b", Forall("z",
        PIff(Mem("z", "b"), POr(PEqv("z", "x"), PEqv("z", "y"))))))),
    # Union
    Forall("a", PExists("b", Forall("x",
        PIff(Mem("x", "b"), PExists("y", PAnd(Mem("y", "a"), Mem("x", "y"))))))),
    # PowerSet
    Forall("a", PExists("b", Forall("x",
        PIff(Mem("x", "b"), Forall("y", Implies(Mem("y", "x"), Mem("y", "a"))))))),
    # Infinity
    PExists("b", PAnd(
        PExists("e", PAnd(Mem("e", "b"), Forall("z", Neg(Mem("z", "e"))))),
        Forall("y", Implies(Mem("y", "b"),
            PExists("s", PAnd(Mem("s", "b"),
                Forall("w", PIff(Mem("w", "s"), POr(Mem("w", "y"), PEqv("w", "y")))))))))),
    # Regularity
    Forall("a", Implies(
        PExists("y", Mem("y", "a")),
        PExists("y", PAnd(Mem("y", "a"), Neg(PExists("z", PAnd(Mem("z", "a"), Mem("z", "y")))))))),
    # Choice
    Forall("x", Implies(
        Forall("y", Implies(Mem("y", "x"), PExists("z", Mem("z", "y")))),
        PExists("c", Forall("y", Implies(Mem("y", "x"),
            PExists("z", PAnd(PAnd(Mem("z", "y"), Mem("z", "c")),
                Forall("w", Implies(PAnd(Mem("w", "y"), Mem("w", "c")), PEqv("w", "z")))))))))),
]

def _strip_foralls(f):
    """Yield (stripped_vars, body) with 0, 1, 2, ... outer foralls stripped."""
    stripped = []
    yield stripped, f
    while isinstance(f, Forall):
        stripped = stripped + [f.var]
        f = f.body
        yield stripped, f

def is_separation(f):
    phi = Capture('phi')
    pat = Forall("a", PExists("b", Forall("x",
        PIff(Mem("x", "b"), PAnd(Mem("x", "a"), phi)))))
    for stripped, g in _strip_foralls(f):
        b = {}
        if pmatch(pat, g, b):
            allowed = set(stripped) | {b["a"], b["x"]}
            if free_vars(b["phi"]) <= allowed:
                return True
    return False

def is_replacement(f):
    phi = Capture('phi')
    uniq = Capture('uniq')
    pat = Forall("a", Implies(uniq, PExists("b", Forall("y",
        PIff(Mem("y", "b"), PExists("x", PAnd(Mem("x", "a"), phi)))))))
    for stripped, g in _strip_foralls(f):
        b = {}
        if pmatch(pat, g, b):
            allowed = set(stripped) | {b["a"], b["x"], b["y"]}
            if free_vars(b["phi"]) <= allowed:
                return True
    return False

# ZFC[0..6] = Ext, Empty, Pair, Union, Power, Inf, Reg
# ZFC[7] = Choice

def is_axiom(f, system):
    # z: no Choice, no Replacement
    # zf: no Choice
    # zfc: all
    limit = 7 if system in ("z", "zf") else 8
    for pattern in ZFC[:limit]:
        if pmatch(pattern, f, {}):
            return True
    if is_separation(f):
        return True
    if system in ("zf", "zfc") and is_replacement(f):
        return True
    return False


# === qed ===

def qed(p, expected, system):
    s = p.sequent
    if len(s.right) != 1:
        raise ValueError(f'qed: expected 1 formula on right, got {len(s.right)}')
    if free_vars(s.right[0]):
        raise ValueError('qed: theorem has free variables')
    for f in s.left:
        if not is_axiom(f, system):
            raise ValueError(f'qed: non-axiom on left (system={system})')
    if not same(s.right[0], expected):
        raise ValueError('qed: theorem does not match expected')

File: poc4/core/kernel.js
"use strict";

// === Types ===

class Mem {
    constructor(left, right) { this.left = left; this.right = right; }
}
class Neg {
    constructor(operand) { this.operand = operand; }
}
class Implies {
    constructor(left, right) { this.left = left; this.right = right; }
}
class Forall {
    constructor(var_, body) { this.var = var_; this.body = body; }
}
class Sequent {
    constructor(left, right) {
        checkSet(left);
        checkSet(right);
        this.left = left;
        this.right = right;
    }
}
class Proof {
    constructor(sequent) { this.sequent = sequent; }
}

function checkSet(lst) {
    for (let i = 0; i < lst.length; i++)
        for (let j = i + 1; j < lst.length; j++)
            if (same(lst[i], lst[j]))
                throw new Error("sequent: duplicate formula");
}


// === Substitution ===

function subst(formula, old, new_) {
    if (typeof formula === "string")
        return formula === old ? new_ : formula;
    if (formula instanceof Mem)
        return new Mem(subst(formula.left, old, new_), subst(formula.right, old, new_));
    if (formula instanceof Neg)
        return new Neg(subst(formula.operand, old, new_));
    if (formula instanceof Implies)
        return new Implies(subst(formula.left, old, new_), subst(formula.right, old, new_));
    if (formula instanceof Forall) {
        if (formula.var === old) return formula;
        return new Forall(formula.var, subst(formula.body, old, new_));
    }
    throw new Error(`cannot subst: ${formula.constructor.name}`);
}


// === Alpha-equivalence ===

function same(a, b) { return eq(a, b, []); }

function eq(a, b, env) {
    if (typeof a === "string" && typeof b === "string") {
        for (const [v1, v2] of env) {
            if (a === v1) return b === v2;
            if (b === v2) return false;
        }
        return a === b;
    }
    if (a.constructor !== b.constructor) return false;
    if (a instanceof Mem)
        return eq(a.left, b.left, env) && eq(a.right, b.right, env);
    if (a instanceof Neg)
        return eq(a.operand, b.operand, env);
    if (a instanceof Implies)
        return eq(a.left, b.left, env) && eq(a.right, b.right, env);
    if (a instanceof Forall)
        return eq(a.body, b.body, [[a.var, b.var], ...env]);
    return false;
}


// === Free variables ===

function freeVars(formula, bound) {
    if (!bound) bound = new Set();
    if (typeof formula === "string")
        return bound.has(formula) ? new Set() : new Set([formula]);
    if (formula instanceof Mem)
        return setUnion(freeVars(formula.left, bound), freeVars(formula.right, bound));
    if (formula instanceof Neg)
        return freeVars(formula.operand, bound);
    if (formula instanceof Implies)
        return setUnion(freeVars(formula.left, bound), freeVars(formula.right, bound));
    if (formula instanceof Forall)
        return freeVars(formula.body, new Set([...bound, formula.var]));
    return new Set();
}

function boundVars(formula) {
    if (typeof formula === "string") return new Set();
    if (formula instanceof Mem)
        return setUnion(boundVars(formula.left), boundVars(formula.right));
    if (formula instanceof Neg) return boundVars(formula.operand);
    if (formula instanceof Implies)
        return setUnion(boundVars(formula.left), boundVars(formula.right));
    if (formula instanceof Forall)
        return new Set([formula.var, ...boundVars(formula.body)]);
    return new Set();
}

function setUnion(a, b) {
    const r = new Set(a);
    for (const x of b) r.add(x);
    return r;
}


// === Helpers ===

function fin(f, lst) { return lst.some(g => same(f, g)); }

function remove(lst, f) {
    let removed = false;
    const result = [];
    for (const g of lst) {
        if (!removed && same(f, g)) { removed = true; }
        else result.push(g);
    }
    return result;
}

function setAdd(lst, f) {
    return fin(f, lst) ? [...lst] : [...lst, f];
}

function eqSequent(a, b) {
    return isPermutation(a.left, b.left) && isPermutation(a.right, b.right);
}

function isPermutation(a, b) {
    if (a.length !== b.length) return false;
    const used = new Array(b.length).fill(false);
    for (const f of a) {
        let found = false;
        for (let j = 0; j < b.length; j++) {
            if (!used[j] && same(f, b[j])) {
                used[j] = true;
                found = true;
                break;
            }
        }
        if (!found) return false;
    }
    return true;
}


// === Proof rules ===

function proof(seq, rule, premises, principal, term) {
    const err = checkRule(seq, rule, premises, principal, term);
    if (err) throw new Error(`invalid proof step: ${rule}\n  ${err}`);
    return new Proof(seq);
}

function checkRule(s, rule, premises, principal, term) {
    const ps = premises.map(p => p.sequent);

    switch (rule) {
        case "axiom":
            if (ps.length !== 0) return "axiom: premises not empty";
            if (!fin(principal, s.left)) return "axiom: principal not in left";
            if (!fin(principal, s.right)) return "axiom: principal not in right";
            return null;

        case "neg_left":
            if (ps.length !== 1 || !(principal instanceof Neg)) return "neg_left: bad principal/premises";
            if (!fin(principal, s.left)) return "neg_left: principal not in left";
            if (!eqSequent(ps[0], new Sequent(remove(s.left, principal), setAdd(s.right, principal.operand))))
                return "neg_left: premise mismatch";
            return null;

        case "neg_right":
            if (ps.length !== 1 || !(principal instanceof Neg)) return "neg_right: bad principal/premises";
            if (!fin(principal, s.right)) return "neg_right: principal not in right";
            if (!eqSequent(ps[0], new Sequent(setAdd(s.left, principal.operand), remove(s.right, principal))))
                return "neg_right: premise mismatch";
            return null;

        case "implies_left": {
            if (ps.length !== 2 || !(principal instanceof Implies)) return "implies_left: bad principal/premises";
            if (!fin(principal, s.left)) return "implies_left: principal not in left";
            const G = remove(s.left, principal);
            if (!eqSequent(ps[0], new Sequent(G, setAdd(s.right, principal.left))))
                return "implies_left: premise 0 mismatch";
            if (!eqSequent(ps[1], new Sequent(setAdd(G, principal.right), s.right)))
                return "implies_left: premise 1 mismatch";
            return null;
        }

        case "implies_right": {
            if (ps.length !== 1 || !(principal instanceof Implies)) return "implies_right: bad principal/premises";
            if (!fin(principal, s.right)) return "implies_right: principal not in right";
            const D = remove(s.right, principal);
            if (!eqSequent(ps[0], new Sequent(setAdd(s.left, principal.left), setAdd(D, principal.right))))
                return "implies_right: premise mismatch";
            return null;
        }

        case "forall_left": {
            if (ps.length !== 1 || typeof term !== "string" || !(principal instanceof Forall))
                return "forall_left: bad principal/premises/term";
            if (!fin(principal, s.left)) return "forall_left: principal not in left";
            const bv = boundVars(principal.body);
            if (bv.has(term)) return `forall_left: term "${term}" clashes with boundVars {${[...bv].join(", ")}}`;
            const G = remove(s.left, principal);
            const substituted = subst(principal.body, principal.var, term);
            if (!eqSequent(ps[0], new Sequent(setAdd(G, substituted), s.right)))
                return "forall_left: premise mismatch";
            return null;
        }

        case "forall_right": {
            if (ps.length !== 1 || typeof term !== "string" || !(principal instanceof Forall))
                return "forall_right: bad principal/premises/term";
            if (!fin(principal, s.right)) return "forall_right: principal not in right";
            const bv = boundVars(principal.body);
            if (bv.has(term)) return `forall_right: term "${term}" clashes with boundVars {${[...bv].join(", ")}}`;
            const D = remove(s.right, principal);
            const all = [...s.left, ...D];
            const fvf = all.find(f => freeVars(f).has(term));
            if (fvf) return `forall_right: term "${term}" free in context`;
            const substituted = subst(principal.body, principal.var, term);
            if (!eqSequent(ps[0], new Sequent(s.left, setAdd(D, substituted))))
                return "forall_right: premise mismatch";
            return null;
        }

        case "cut":
            if (ps.length !== 2) return "cut: need 2 premises";
            if (!eqSequent(ps[0], new Sequent(s.left, setAdd(s.right, principal))))
                return "cut: premise 0 mismatch";
            if (!eqSequent(ps[1], new Sequent(setAdd(s.left, principal), s.right)))
                return "cut: premise 1 mismatch";
            return null;

        case "weakening_left": {
            if (ps.length !== 1) return "weakening_left: need 1 premise";
            if (!fin(principal, s.left)) return "weakening_left: principal not in left";
            if (fin(principal, ps[0].left)) {
                if (!eqSequent(ps[0], s)) return "weakening_left: premise mismatch (already present)";
                return null;
            }
            if (!eqSequent(ps[0], new Sequent(remove(s.left, principal), s.right)))
                return "weakening_left: premise mismatch";
            return null;
        }

        case "weakening_right": {
            if (ps.length !== 1) return "weakening_right: need 1 premise";
            if (!fin(principal, s.right)) return "weakening_right: principal not in right";
            if (fin(principal, ps[0].right)) {
                if (!eqSequent(ps[0], s)) return "weakening_right: premise mismatch (already present)";
                return null;
            }
            if (!eqSequent(ps[0], new Sequent(s.left, remove(s.right, principal))))
                return "weakening_right: premise mismatch";
            return null;
        }
    }
    return `unknown rule: ${rule}`;
}


// === ZFC pattern matching ===

class Capture {
    constructor(name) { this.name = name; }
}

const ANY = new Capture("_");

function pmatch(pattern, formula, bindings) {
    if (!bindings) bindings = {};
    if (pattern instanceof Capture) {
        bindings[pattern.name] = formula;
        return true;
    }
    if (typeof pattern === "string" && typeof formula === "string") {
        if (pattern in bindings) return bindings[pattern] === formula;
        bindings[pattern] = formula;
        return true;
    }
    if (pattern.constructor !== formula.constructor) return false;
    if (pattern instanceof Mem)
        return pmatch(pattern.left, formula.left, bindings) && pmatch(pattern.right, formula.right, bindings);
    if (pattern instanceof Neg)
        return pmatch(pattern.operand, formula.operand, bindings);
    if (pattern instanceof Implies)
        return pmatch(pattern.left, formula.left, bindings) && pmatch(pattern.right, formula.right, bindings);
    if (pattern instanceof Forall)
        return pmatch(pattern.var, formula.var, bindings) && pmatch(pattern.body, formula.body, bindings);
    return false;
}

// Derived helpers for building patterns
function PAnd(a, b) { return new Neg(new Implies(a, new Neg(b))); }
function POr(a, b) { return new Implies(new Neg(a), b); }
function PIff(a, b) { return new Neg(new Implies(new Implies(a, b), new Neg(new Implies(b, a)))); }
function PExists(v, body) { return new Neg(new Forall(v, new Neg(body))); }
function PEqv(a, b, z) { z = z || "_z"; return new Forall(z, PIff(new Mem(z, a), new Mem(z, b))); }

const ZFC = [
    // Extensionality
    new Forall("x", new Forall("y", new Implies(
        new Forall("z", PIff(new Mem("z", "x"), new Mem("z", "y"))),
        new Forall("z", PIff(new Mem("x", "z"), new Mem("y", "z")))))),
    // EmptySet
    PExists("b", new Forall("x", new Neg(new Mem("x", "b")))),
    // Pairing
    new Forall("x", new Forall("y", PExists("b", new Forall("z",
        PIff(new Mem("z", "b"), POr(PEqv("z", "x"), PEqv("z", "y"))))))),
    // Union
    new Forall("a", PExists("b", new Forall("x",
        PIff(new Mem("x", "b"), PExists("y", PAnd(new Mem("y", "a"), new Mem("x", "y"))))))),
    // PowerSet
    new Forall("a", PExists("b", new Forall("x",
        PIff(new Mem("x", "b"), new Forall("y", new Implies(new Mem("y", "x"), new Mem("y", "a"))))))),
    // Infinity
    PExists("b", PAnd(
        PExists("e", PAnd(new Mem("e", "b"), new Forall("z", new Neg(new Mem("z", "e"))))),
        new Forall("y", new Implies(new Mem("y", "b"),
            PExists("s", PAnd(new Mem("s", "b"),
                new Forall("w", PIff(new Mem("w", "s"), POr(new Mem("w", "y"), PEqv("w", "y")))))))))),
    // Regularity
    new Forall("a", new Implies(
        PExists("y", new Mem("y", "a")),
        PExists("y", PAnd(new Mem("y", "a"), new Neg(PExists("z", PAnd(new Mem("z", "a"), new Mem("z", "y")))))))),
    // Choice
    new Forall("x", new Implies(
        new Forall("y", new Implies(new Mem("y", "x"), PExists("z", new Mem("z", "y")))),
        PExists("c", new Forall("y", new Implies(new Mem("y", "x"),
            PExists("z", PAnd(PAnd(new Mem("z", "y"), new Mem("z", "c")),
                new Forall("w", new Implies(PAnd(new Mem("w", "y"), new Mem("w", "c")), PEqv("w", "z")))))))))),
];

function* stripForalls(f) {
    let stripped = [];
    yield [stripped, f];
    while (f instanceof Forall) {
        stripped = [...stripped, f.var];
        f = f.body;
        yield [stripped, f];
    }
}

function isSeparation(f) {
    const phi = new Capture("phi");
    const pat = new Forall("a", PExists("b", new Forall("x",
        PIff(new Mem("x", "b"), PAnd(new Mem("x", "a"), phi)))));
    for (const [stripped, g] of stripForalls(f)) {
        const b = {};
        if (pmatch(pat, g, b)) {
            const allowed = new Set([...stripped, b["a"], b["x"]]);
            if (isSubset(freeVars(b["phi"]), allowed)) return true;
        }
    }
    return false;
}

function isReplacement(f) {
    const phi = new Capture("phi");
    const uniq = new Capture("uniq");
    const pat = new Forall("a", new Implies(uniq, PExists("b", new Forall("y",
        PIff(new Mem("y", "b"), PExists("x", PAnd(new Mem("x", "a"), phi)))))));
    for (const [stripped, g] of stripForalls(f)) {
        const b = {};
        if (pmatch(pat, g, b)) {
            const allowed = new Set([...stripped, b["a"], b["x"], b["y"]]);
            if (isSubset(freeVars(b["phi"]), allowed)) return true;
        }
    }
    return false;
}

function isSubset(a, b) {
    for (const x of a) if (!b.has(x)) return false;
    return true;
}

function isAxiom(f, system) {
    const limit = (system === "z" || system === "zf") ? 7 : 8;
    for (let i = 0; i < limit; i++)
        if (pmatch(ZFC[i], f, {})) return true;
    if (isSeparation(f)) return true;
    if ((system === "zf" || system === "zfc") && isReplacement(f)) return true;
    return false;
}


// === qed ===

function qed(p, expected, system) {
    const s = p.sequent;
    if (s.right.length !== 1)
        throw new Error(`qed: expected 1 formula on right, got ${s.right.length}`);
    if (freeVars(s.right[0]).size > 0)
        throw new Error("qed: theorem has free variables");
    for (const f of s.left)
        if (!isAxiom(f, system))
            throw new Error(`qed: non-axiom on left (system=${system})`);
    if (!same(s.right[0], expected))
        throw new Error("qed: theorem does not match expected");
}


module.exports = {
    Mem, Neg, Implies, Forall, Sequent, Proof, Capture, ANY,
    subst, same, freeVars, boundVars, fin, remove, setAdd,
    proof, qed, pmatch, PAnd, POr, PIff, PExists, PEqv,
    isAxiom, isSeparation, isReplacement,
};
